pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/BinaryWorkloadScheduler.java

```diff
@@ -0,0 +1,212 @@
+import org.apache.pinot.core.query.scheduler.resources.BinaryWorkloadResourceManager;
+import org.apache.pinot.core.query.scheduler.resources.QueryExecutorService;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This scheduler is designed to deal with two types of workloads
+ * 1. Primary Workloads -> regular queries from the application
+ * 2. Secondary Workloads -> adhoc queries fired from tools, testing, etc
+ *
+ *
+ * Primary Workload Queries
+ * Primary workloads queries are executed with priority and submitted to the Runner threads as and when they arrive.
+ * The resources used by a primary workload query is not capped.
+ *
+ * Secondary Workload Queries
+ *   - Secondary workload queries are identified using a query option -> "SET isSecondaryWorkload=true"
+ *   - Secondary workload queries are contained as follows:
+ *       - Restrictions on number of runner threads available to process secondary queries
+ *       - Restrictions on total number of worker threads available to process a single secondary query
+ *       - Restrictions on total number of worker threads available to process all in-progress secondary queries
+ */
+public class BinaryWorkloadScheduler extends QueryScheduler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BinaryWorkloadScheduler.class);
+
+  public static final String MAX_SECONDARY_QUERIES = "binarywlm.maxSecondaryRunnerThreads";
+  public static final int DEFAULT_MAX_SECONDARY_QUERIES = 5;
+
+  // Secondary Workload Runners.
+  private final int _numSecondaryRunners;
+  private final Semaphore _secondaryRunnerSemaphore;
+
+  private final SecondaryWorkloadQueue _secondaryQueryQ;
+
+  Thread _scheduler;
+
+  public BinaryWorkloadScheduler(PinotConfiguration config, QueryExecutor queryExecutor, ServerMetrics metrics,
+      LongAccumulator latestQueryTime) {
+    super(config, queryExecutor, new BinaryWorkloadResourceManager(config), metrics, latestQueryTime);
+
+    _secondaryQueryQ = new SecondaryWorkloadQueue(config, _resourceManager);
+    _numSecondaryRunners = config.getProperty(MAX_SECONDARY_QUERIES, DEFAULT_MAX_SECONDARY_QUERIES);
+    LOGGER.info("numSecondaryRunners={}", _numSecondaryRunners);
+    _secondaryRunnerSemaphore = new Semaphore(_numSecondaryRunners);
+  }
+
+  @Override
+  public String name() {
+    return "BinaryWorkloadScheduler";
+  }
+
+  @Override
+  public ListenableFuture<byte[]> submit(ServerQueryRequest queryRequest) {
+    if (!_isRunning) {
+      return immediateErrorResponse(queryRequest, QueryException.SERVER_SCHEDULER_DOWN_ERROR);
+    }
+
+    queryRequest.getTimerContext().startNewPhaseTimer(ServerQueryPhase.SCHEDULER_WAIT);
+    if (!QueryOptionsUtils.isSecondaryWorkload(queryRequest.getQueryContext().getQueryOptions())) {
+      QueryExecutorService queryExecutorService = _resourceManager.getExecutorService(queryRequest, null);
```

Comment:
@vvivekiyer : can you elaborate on the decision to make the primary workload use Unbounded FCFS?

Given this scheduler's goal is to prevent cluster degradation for tables/tenants with multiple use-cases on the same table, wouldn't it make sense to support scheduler groups for the primary workloads as well?

Reply:
For Primary workloads, we wanted to retain existing default behavior without any disruption (as they correspond to all our production queries) - hence the unbounded FCFS.  Using an unbounded FCFS for primary ensures that Primary workloads can always eat into the resources of secondary workloads but not vice-versa.